<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>Stdlib__fun (ocaml.Stdlib__fun)</title><link rel="stylesheet" href="../../_odoc-theme/odoc.css"/><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0"/><script src="../../highlight.pack.js"></script><script>hljs.initHighlightingOnLoad();</script></head><body><div class="content"><header><nav><a href="../index.html">Up</a> â€“ <a href="../index.html">ocaml</a> &#x00BB; Stdlib__fun</nav><h1>Module <code>Stdlib__fun</code></h1><p>Function manipulation.</p><dl><dt>since</dt><dd>4.08</dd></dl><nav class="toc"><ul><li><a href="#combinators">Combinators</a></li><li><a href="#exception-handling">Exception handling</a></li></ul></nav></header><section><header><h2 id="combinators"><a href="#combinators" class="anchor"></a>Combinators</h2></header><dl><dt class="spec external" id="val-id"><a href="#val-id" class="anchor"></a><code><span class="keyword">val</span> id : <span class="type-var">'a</span> <span>&#45;&gt;</span> <span class="type-var">'a</span></code></dt><dd><p><code>id</code> is the identity function. For any argument <code>x</code>, <code>id x</code> is <code>x</code>.</p></dd></dl><dl><dt class="spec value" id="val-const"><a href="#val-const" class="anchor"></a><code><span class="keyword">val</span> const : <span class="type-var">'a</span> <span>&#45;&gt;</span> <span class="type-var">_</span> <span>&#45;&gt;</span> <span class="type-var">'a</span></code></dt><dd><p><code>const c</code> is a function that always returns the value <code>c</code>. For any argument <code>x</code>, <code>(const c) x</code> is <code>c</code>.</p></dd></dl><dl><dt class="spec value" id="val-flip"><a href="#val-flip" class="anchor"></a><code><span class="keyword">val</span> flip : (<span class="type-var">'a</span> <span>&#45;&gt;</span> <span class="type-var">'b</span> <span>&#45;&gt;</span> <span class="type-var">'c</span>) <span>&#45;&gt;</span> <span class="type-var">'b</span> <span>&#45;&gt;</span> <span class="type-var">'a</span> <span>&#45;&gt;</span> <span class="type-var">'c</span></code></dt><dd><p><code>flip f</code> reverses the argument order of the binary function <code>f</code>. For any arguments <code>x</code> and <code>y</code>, <code>(flip f) x y</code> is <code>f y x</code>.</p></dd></dl><dl><dt class="spec value" id="val-negate"><a href="#val-negate" class="anchor"></a><code><span class="keyword">val</span> negate : (<span class="type-var">'a</span> <span>&#45;&gt;</span> bool) <span>&#45;&gt;</span> <span class="type-var">'a</span> <span>&#45;&gt;</span> bool</code></dt><dd><p><code>negate p</code> is the negation of the predicate function <code>p</code>. For any argument <code>x</code>, <code>(negate p) x</code> is <code>not (p x)</code>.</p></dd></dl></section><section><header><h2 id="exception-handling"><a href="#exception-handling" class="anchor"></a>Exception handling</h2></header><dl><dt class="spec value" id="val-protect"><a href="#val-protect" class="anchor"></a><code><span class="keyword">val</span> protect : finally:(unit <span>&#45;&gt;</span> unit) <span>&#45;&gt;</span> (unit <span>&#45;&gt;</span> <span class="type-var">'a</span>) <span>&#45;&gt;</span> <span class="type-var">'a</span></code></dt><dd><p><code>protect ~finally work</code> invokes <code>work ()</code> and then <code>finally ()</code> before <code>work ()</code> returns with its value or an exception. In the latter case the exception is re-raised after <code>finally ()</code>. If <code>finally ()</code> raises an exception, then the exception <a href="index.html#exception-Finally_raised"><code>Finally_raised</code></a> is raised instead.</p><p><code>protect</code> can be used to enforce local invariants whether <code>work
    ()</code> returns normally or raises an exception. However, it does not protect against unexpected exceptions raised inside <code>finally ()</code> such as <a href="../Stdlib/index.html#exception-Out_of_memory"><code>Stdlib.Out_of_memory</code></a>, <a href="../Stdlib/index.html#exception-Stack_overflow"><code>Stdlib.Stack_overflow</code></a>, or asynchronous exceptions raised by signal handlers (e.g. <span class="xref-unresolved" title="unresolved reference to &quot;Sys.Break&quot;"><code>Sys</code>.Break</span>).</p><p>Note: It is a <em>programming error</em> if other kinds of exceptions are raised by <code>finally</code>, as any exception raised in <code>work ()</code> will be lost in the event of a <a href="index.html#exception-Finally_raised"><code>Finally_raised</code></a> exception. Therefore, one should make sure to handle those inside the finally.</p></dd></dl><dl><dt class="spec exception" id="exception-Finally_raised"><a href="#exception-Finally_raised" class="anchor"></a><code><span class="keyword">exception</span> </code><code><span class="exception">Finally_raised</span> <span class="keyword">of</span> exn</code></dt><dd><p><code>Finally_raised exn</code> is raised by <code>protect ~finally work</code> when <code>finally</code> raises an exception <code>exn</code>. This exception denotes either an unexpected exception or a programming error. As a general rule, one should not catch a <code>Finally_raised</code> exception except as part of a catch-all handler.</p></dd></dl></section></div></body></html>